File: tokicontest/rekonstruksiarray.py
def main():
	n, m = input().split()
	n = int(n)
	m = int(m)
	lst = input().split()
	lst = [int(x) for x in lst]
	counter = n%m
	dicti1 = {}
	dicti2 = {}
	for i in range (1, n+1):
		if i%m not in dicti1:
			dicti1[i%m] = [i]
		else:
			dicti1[i%m] += [i]
	for i in range (m):
		if i == 0:
			dicti2[i] = n//m
		else:
			if counter > 0:
				dicti2[i] = (n//m) + 1
				counter -= 1
			else:
				dicti2[i] = n//m
	saver = {}
	for i in range (len(lst)):
		if lst[i] not in saver:
			saver[lst[i]] = 1
		else:
			saver[lst[i]] += 1
	if dicti2 != saver:
		print(-1)
	else:
		isfirst = True
		for i in range (len(lst)):
			if i != len(lst)-1:
				temp = list(dicti1[lst[i]])
				x = temp.pop()
				dicti1[lst[i]] = temp
				print(x, end=" ")
			else:
				temp = list(dicti1[lst[i]])
				x = temp.pop()
				dicti1[lst[i]] = temp
				print(x)



if __name__ == '__main__':
	main()

File: tokicontest/rekonstruksiarray.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	int n, m;
	cin >> n >> m;
	vector<int> lst;
	for (int i = 0; i < n; i++) {
		int elem;
		cin >> elem;
		lst.push_back(elem);
	}
	int counter = n%m;
	unordered_map<int, stack<int>> dicti1;
	unordered_map<int, int> dicti2;
	unordered_map<int, int> saver;
	for (int i = 1; i < n+1; i++) {
		dicti1[i%m].push(i);
		dicti2[i%m]++;
		saver[lst[i-1]]++;
	}
	if (dicti2 != saver) {
		cout << -1 << endl;
	}
	else {
		bool isfirst = true;
		for (int i = 0; i < lst.size(); i++) {
			if (isfirst) {
				if (dicti1[lst[i]].size() > 0) {
					cout << dicti1[lst[i]].top();
					dicti1[lst[i]].pop();
					isfirst = false;
				}
			}
			else {
				if (dicti1[lst[i]].size() > 0) {
					cout << ' ' << dicti1[lst[i]].top();
					dicti1[lst[i]].pop();
				}
			}
		}
		cout << endl;
	}
}